book/src/read-only-read-write-accounts.md

```diff
@@ -0,0 +1,136 @@
+
+* Read-only access to userdata.
+
+Instructions that debit or modify the read only account userdata will fail.
+
+## Account Lock Optimizations
+
+The Accounts module that keeps track of current locked accounts in the runtime
+separates read-only accounts from the read-write accounts.  The read-only
+accounts can be cached in memory and shared between all the threads executing
+transactions.
+
+The current runtime can't predict if an account is read-only or read-write when
+the transaction account keys are locked at the start of the transaction
+processing pipeline.  Accounts referenced by the transaction have not been
+loaded from disk yet.
+
+An ideal design would cache the read-only accounts while they are referenced by
+any transaction moving through the runtime, and release the cache when the last
+transaction exits the runtime.
+
+## Transaction changes
+
+To enable the possibility of caching accounts only while they are in the
+runtime, the Transaction structure should be changed in the following way:
+
+* `program_ids: Vec<Pubkey>` - This vector is removed.  Program keys can be
+placed at the end of the `account_keys` vector with the `read_only_accounts`
+number set to the right number of programs.
+
+* `read_only_accounts: u8` - The number of keys from the **end** of the
+Transaction's `account_keys` array that are read-only.
+
+The following possible accounts present in an transaction:
+
+* paying account
+* RW accounts
+* RO accounts
+* Program ids
+
+The paying account must be read-write, and program ids must be read-only.  The
+first account in the `account_keys` array is always the account that pays for
+the transaction fee, therefore it cannot be read-only.  For these reasons the
+read-only accounts are all grouped together at the end of the `account_keys`
+vector.  Counting read-only accounts from the end allows for the default `0`
+value to still be functionally correct, since a transaction will succeed with
+all read-write accounts.
+
+Since accounts can only appear once in the transaction's `account_keys` array,
+an account can only be read-only or read-write in a single transaction, not
+both.  The runtime treats a transaction as one atomic unit of execution. If any
+instruction needs read-write access to an Account a copy needs to be made the
+write lock needs to be held for the entire time the transaction is being
+processed by the runtime the runtime.
+
+## Starvation
+
+Read locks can keep the runtime from executing transactions requesting a write.
+
+When a request for a write lock is made while a read lock is open, the
+transaction requesting the write lock should be cached.  Upon closing the read
+lock the pending transactions can be pushed through the runtime.
+
+While a pending write transaction exists, any additional read lock requests for
+that account should fail.  It follows that any other write lock requests will
+fail as well.  Currently clients must retransmit when the transaction fails due
```

Comment:
retransmit here really means (1) getting a new `last_id`, (2) resigning, and *then* (3) retransmit right?

Reply:
@mvines last_id's last two minutes, right?  A client would probably just ask for the finalized root last_id and use that one for a while.